File: Shared/NearbyTrain.swift
//
//  NearbyTrain.swift
//  Now Departing
//
//  Created by Jonathan Bobrow on 6/4/25.
//

import Foundation
import SwiftUI

struct NearbyTrain: Identifiable, Equatable {
    let id = UUID()
    let lineId: String
    let stationId: String
    let stationName: String
    let stationDisplay: String
    let direction: String
    let destination: String
    let arrivalTime: Date  // Store actual arrival time instead of minutes
    let distanceInMeters: Double
    
    // Helper computed properties
    var minutes: Int {
        let timeInterval = arrivalTime.timeIntervalSinceNow
        return max(0, Int(timeInterval / 60))
    }
    
    var timeText: String {
        let timeInterval = arrivalTime.timeIntervalSinceNow
        let totalSeconds = max(0, Int(timeInterval))
        let minutes = totalSeconds / 60
        
        if totalSeconds <= 0 {
            return "Now"
        } else {
            return "\(minutes)m"
        }
    }
    
    // Get live time text with current time for precise countdown
    func getLiveTimeText(currentTime: Date, fullText: Bool = false) -> String {
        let timeInterval = arrivalTime.timeIntervalSince(currentTime)
        let totalSeconds = max(0, Int(timeInterval))
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        
        if totalSeconds <= 30 {
            return "Now"
        } else if totalSeconds < 60 {
            return "Soon"
        } else {
            if(fullText) {
                return "\(minutes) min"
            }
            else {
                return "\(minutes)m"
            }
        }
    }
    
    var distanceText: String {
        // Convert meters to feet
        let feet = distanceInMeters * 3.28084
        
        if feet < 1000 {
            return "\(Int(feet))ft"
        } else {
            // Convert to miles (5280 feet = 1 mile)
            let miles = feet / 5280
            if miles < 10 {
                return String(format: "%.1fmi", miles)
            } else {
                return "\(Int(miles))mi"
            }
        }
    }
    
    var directionText: String {
        return direction == "N" ? "Uptown" : "Downtown"
    }
}

File: NowDepartingAndroid/app/src/main/java/com/move38/nowdeparting/data/model/NearbyTrain.kt
package com.move38.nowdeparting.data.model

import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.UUID

data class NearbyTrain(
    val id: String = UUID.randomUUID().toString(),
    val lineId: String,
    val stationId: String,
    val stationName: String,
    val stationDisplay: String,
    val direction: String,
    val destination: String,
    val arrivalTime: Instant,
    val distanceInMeters: Double
) {
    val minutes: Long get() {
        val now = Instant.now()
        val diffSeconds = ChronoUnit.SECONDS.between(now, arrivalTime)
        return diffSeconds / 60
    }

    val seconds: Long get() {
        val now = Instant.now()
        return ChronoUnit.SECONDS.between(now, arrivalTime)
    }

    val timeText: String get() {
        val mins = minutes
        return when {
            mins < 0 -> "--"
            mins == 0L -> "Now"
            mins == 1L -> "1 min"
            else -> "$mins min"
        }
    }

    val preciseTimeText: String get() {
        val secs = seconds
        return when {
            secs < 0 -> "--"
            secs < 30 -> "Now"
            secs < 60 -> "${secs}s"
            else -> {
                val mins = secs / 60
                if (mins == 1L) "1 min" else "$mins min"
            }
        }
    }

    val distanceText: String get() {
        val feet = distanceInMeters * 3.28084
        return if (feet < 1000) {
            "${feet.toInt()} ft"
        } else {
            val miles = feet / 5280
            String.format("%.1f mi", miles)
        }
    }

    val directionText: String get() {
        return when (direction) {
            "N" -> "Uptown"
            "S" -> "Downtown"
            else -> direction
        }
    }
}
